File: cam5.cpp
#include <cstdio>
#include <vector>
#include <stack>

using namespace std;

class Graph
{
private:
	int v;
	vector<int> *list;
public:
	Graph(int n){
		v = n;
		list = new vector<int>[n];
	}
	void add_edge(int a, int b)
	{
		list[a].push_back(b);
		list[b].push_back(a);
	}
	void dfs(int source, vector<bool> &visited)
	{
		if(visited[source])
			return;

		stack<int> st;
		st.push(source);
		while(!st.empty())
		{
			int top = st.top(); st.pop();
			visited[top] = true;
			for(int i=0;i<list[top].size();i++)
			{
				if(!visited[list[top][i]])
				{
					st.push(list[top][i]);
				}
			}
		}
	}

	int num_connected_components()
	{
		vector<bool> visited(v, false);
		int ans = 0;
		for(int i=0;i<v;i++)
		{
			if(!visited[i])
			{
				ans++;
				dfs(i, visited);
			}
		}
		return ans;
	}
};
int main()
{
	Graph g(4);
	int t,n,a,b,e;
	for(scanf("%d",&t);t--;)
	{
		scanf("%d",&n);
		Graph g(n);
		scanf("%d",&e);
		for(int i=0;i<e;i++)
		{
			scanf("%d %d",&a,&b);
			g.add_edge(a,b);
		}
		printf("%d\n", g.num_connected_components());
	}
}

File: old/mixed/python/cam5.py
class Graph:
	def __init__(self, n):
		self.n = n
		self.list = [[] for i in range(0,n)]

	def add_edge(self, a, b):
		self.list[a].append(b)
		self.list[b].append(a)

	def dfs(self, s, visited):
		st = []
		if not visited[s]:
			st.append(s)
			while len(st)!=0:
				top = st.pop()
				visited[top]=True
				for nb in self.list[top]:
					if not visited[nb]:
						visited[nb]=True
						st.append(nb)

	def num_connected_components(self):
		visited = [ False for i in range(0, self.n)]
		ans = 0
		for i in range(0, self.n):
			if not visited[i]:
				ans += 1
				self.dfs(i, visited)
		return ans

t = int(input())
# x = input()
while t>0:
	n = int(input())
	g = Graph(n)
	e = int(input())
	for _ in range(e):
		x = input().strip().split()
		a = int(x[0])
		b = int(x[1])
		g.add_edge(a, b)
	print(g.num_connected_components())
	t = t-1
	# x = input()